### ISSUE TITLE
Regular Expressions in Speech Dictionaries don't support Word Boundary

### ISSUE THREAD

**Issue by school510587** [nvaccess/nvda#6830]
Some Chinese users would like to hear continuous Chinese words when reading text consisting of Chinese and English words.
For example, 我的dog是黑色的 (my dog is black). "dog" is equivalent to 小狗 in Chinese, so we want to hear 我的小狗是黑色的.
There is currently one class "Whole word" in speech dictionaries, but it requires spaces around "dog". In other words, it doesn't work on 我的dog是黑色的, and it works on 我的  dog  是黑色的. However, most English words have no surrounding spaces.
It is another solution to use regular expression \bdog\b , but I find it doesn't work, either.
After some advanced experiments, I find it matches the string '\x08dog\x08'.
\b is treated as \x08 rather than word boundaries, which is its original meaning in regular expressions.
PS. I did set the correct class "regular expression" for my speech dictionary rule.

**Comment by jcsteh** [comment]
1. The whole word option just uses \b internally. It would seem that the regular expression entry doesn't see "dog" as a separate word here. That suggests that these Chinese characters aren't considered by the Unicode database to be word boundaries.
2. While \b will indeed match \x08, that's not because it's being incorrectly translated to \x08. This occurs because \x08 is not considered to be a word character. As for why it doesn't achieve what you want, the reason is the same as above.

There's nothing we can do to fix the regular expression \b command. If there's some other way to match whole words for Chinese that doesn't break other languages, we could use that for the whole word option. However, I don't have any idea what that would be.

**Comment by jcsteh** [comment]
I confirmed this was a problem with the \b match using Python as follows:

```
# First, without spaces:
>>> re.compile(ur"\bdog\b", re.UNICODE).search(u"我的dog是黑色的")
# Got None.
# Now, with spaces:
>>> re.compile(ur"\bdog\b", re.UNICODE).search(u"我的 dog 是黑色的")
<_sre.SRE_Match object at 0x05B2A448>
```

I confirmed that \b is not being treated as backspace as follows:

1. NVDA menu -> Preferences -> Speech dictionaries -> Temporary dictionary.
2. Add.
3. Pattern: \bdog\b
4. Replacement: cat
5. Type: Regular expression
6. OK, OK.
7. Open Notepad, paste this text and then read it: 我的 dog 是黑色的
 * Result: 我的 cat 是黑色的
 * Note that the replacement occurred with spaces (not \x08 characters).

**Comment by school510587** [comment]
@jcsteh, thanks for your reply.
OK, I tried re.search without re.UNICODE flag by mistake yesterday. So "Whole word" acts as expected.
However, I tested the second problem by print("\bdog\b") to a txt file and read it by NVDA.
Following your setting steps, NVDA said "cat" when the cursor moved to the line containing only "dog" surrounded by two backspaces.

**Comment by jcsteh** [comment]
I'm not sure that explaining the effect of Unicode will solve the confusion
for users. Without Unicode, Unicode characters that are undoubtedly word
boundaries (e.g. full width white space) won't be treated as word
boundaries, even though \b says it relates to word boundaries. The only
reason that \b works without Unicode in your case is that you're only
considering ASCII word boundaries, so Chinese characters don't count as
word characters. Surely, that's not usually what Chinese users want; i.e.
Chinese characters should be considered as word characters in most cases.

**Comment by school510587** [comment]
I can't ensure that no user will get confused in the future, too. However, Unicode flag is not raised by default in most circumstances.
Moreover, basic regex tutorials don't mention complex effects after raising Unicode flag, which will let users don't know why their regexes didn't work.
When a regex don't work, users' first action may be consulting NVDA's user guide and their basic regex tutorial, and they find nothing useful to identify their mistakes.
Thus, I think user guide can help us by WARNING us that NVDA's regex runs with Unicode flag raised.

**Comment by jcsteh** [comment]
I would have thought basic regex tutorials would just talk about word
boundaries, rather than specifically mentioning what characters are
covered. If I'm correct (and please indicate if I'm not), why would
informing the user about the Unicode flag help?

**Comment by school510587** [comment]
Many regex tutorials in *Chinese* still present the behaviors of regexes in ASCII domain without mentioning Unicode.
Those tutorials even state that \b matches the boundaries of "English words".
I know that regex tutorials in English are often very comprehensive.
However, users may not automatically consult English versions of tutorials to find their problems.
Also, I think that we can't expect each user will test everything in Python, just like what you did in Python codes above.
Considering my original purpose, I wanted to replace *English* words in text containing Chinese and English characters with some Chinese terms.
Many Chinese users, including me, don't think that one "word" will consist of consecutive Chinese and English characters. We think they are two words regardless of the number of spaces between them, because Chinese and English are two different sets of characters.
After reading those Chinese tutorials, users think that Chinese characters are "boundaries of English words", and they don't know why NVDA's regex replacement function doesn't work for them.
In fact, many users have no idea about Unicode, and they won't realize that their concepts of regexes are not complete.
If NVDA's user guide gives a hint that users should consider regexes under Unicode domain, then users will be aware of what they are missing.

**Comment by feerrenrut** [comment]
@school510587 do you think that you could propose a modification to the user guide for this?

**Comment by school510587** [comment]
I have read the English version of userGuide.t2t:

Setting the pattern to match as a whole word means that the replacement will only be made if the pattern does not occur as part of a larger word; i.e. a character other than an alphanumeric character or an underscore (or no character at all) comes both immediately before and after the pattern.

This statement is still under ASCII domain. Word characters contain more than just alphanumeric characters and an underscore.
If it was me, I will propose to change this line to:

Setting the pattern to match as a whole word means that the replacement will only be made if the pattern does not occur as part of a larger word; i.e. a non-word character (or no character at all) comes both immediately before and after the pattern.
Note: A word character may be an alphanumeric character, an underscore, a Han or Kana character etc.

(Sorry, please forgive my poor English. Perhaps there exists better words to describe this.)
Also, add a note after this line:

 Regular expressions are not covered in this user guide, but there are many tutorials on the web which can provide you with more information.
+Note: When this functionality doesn't work with your regular expressions, please consult a more comprehensive reference of regular expression mentioning Unicode.

**Comment by jcsteh** [comment]
The problem with the term non-word character is that it is extremely
confusing in this context. I seem to recall we actually used to say
non-word character here, but this was confusing for people.

To clarify, are you saying you don't consider Chinese letters and numbers
to be alphanumeric? Alphanumeric is just defined as a character which is
considered to be part of an alphabet or a character which is a number,
regardless of language. I guess Chinese characters can't really be
considered part of an alphabet, since they're pictographic. What if we said
letters, numbers or underscore? That eliminates the term alphabet, while
still keeping it language neutral. "Letter" might still not be quite the
right term, though.

**Comment by school510587** [comment]
Yes, Chinese characters are in \w but are not in [:alpha:].
I think "letter" is OK, because I can't figure out simpler way to describe \w or \b, either.
Maybe other user will provide some better words in English to clarify it.
In fact, the translation of "letter" in Chinese is similar to "word" in this context. If there is still serious ambiguity in the translated user guide in Chinese, I will propose something to the l10n team of my locale to improve it.
After all, the two places of user guide is related to this discussion, and I hope that text there to be adjusted. Thanks!

**Comment by XLTechie** [comment]
@school510587 is this still a problem in 2023 under Python 3 re module, and what is the status of your attempts to arrange a solution that is satisfactory?
Put another way: why is this still open? Is some action pending?

**Comment by school510587** [comment]
Hi @jcsteh and @XLTechie,

I have discussed the issue with the zh_TW l10n team. The result of program execution has no problem. The source of the issue is some different comprehension of "a whole word" between English users and Chinese users. Thus, we have solved this issue by an improvement of the user guide. The issue can be closed when the following commit is included into the NVDA main repository:

```
71561; Actions: Action column;
Author: wangjanli;
Date: 2023年01月20日 下午 10:49:22;
Message: Soved GitHub issue #6830.
```

Thanks!
